cpp/examples/arrow/execution_plan_documentation_examples.cc

```diff
@@ -0,0 +1,1160 @@
+
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+    cp::ExecNode* sink;
+
+    ARROW_ASSIGN_OR_RAISE(
+        sink, cp::MakeExecNode("sink", plan.get(), {scan},
+        cp::SinkNodeOptions{&sink_gen}));
+
+    // // translate sink_gen (async) to sink_reader (sync)
+    std::shared_ptr<arrow::RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+        dataset->schema(), std::move(sink_gen), exec_context.memory_pool());
+
+    // validate the ExecPlan
+    ABORT_ON_FAILURE(plan->Validate());
+    PRINT_LINE("ExecPlan created : " << plan->ToString());
+    // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+
+    // // collect sink_reader into a Table
+    std::shared_ptr<arrow::Table> response_table;
+
+    ARROW_ASSIGN_OR_RAISE(response_table,
+                            arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+    PRINT_LINE("Results : " << response_table->ToString());
+
+    // // stop producing
+    plan->StopProducing();
+    // // plan mark finished
+    plan->finished().Wait();
+
+    return arrow::Status::OK();
+}
+
+cp::ExecBatch GetExecBatchFromJSON(const std::vector<arrow::ValueDescr>& descrs,
+                                   arrow::util::string_view json) {
+  auto fields = ::arrow::internal::MapVector(
+      [](const arrow::ValueDescr& descr) { return arrow::field("", descr.type); },
+      descrs);
+
+  cp::ExecBatch batch{*GetRecordBatchFromJSON(arrow::schema(std::move(fields)), json)};
+
+  auto value_it = batch.values.begin();
+  for (const auto& descr : descrs) {
+    if (descr.shape == arrow::ValueDescr::SCALAR) {
+      if (batch.length == 0) {
+        *value_it = arrow::MakeNullScalar(value_it->type());
+      } else {
+        *value_it = value_it->make_array()->GetScalar(0).ValueOrDie();
+      }
+    }
+    ++value_it;
+  }
+
+  return batch;
+}
+
+struct BatchesWithSchema {
+  std::vector<cp::ExecBatch> batches;
+  std::shared_ptr<arrow::Schema> schema;
+
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> gen(bool parallel) const {
+    auto opt_batches = ::arrow::internal::MapVector(
+        [](cp::ExecBatch batch) { return arrow::util::make_optional(std::move(batch)); },
+        batches);
+
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> gen;
+
+    if (parallel) {
+      // emulate batches completing initial decode-after-scan on a cpu thread
+      gen = arrow::MakeBackgroundGenerator(
+                arrow::MakeVectorIterator(std::move(opt_batches)),
+                ::arrow::internal::GetCpuThreadPool())
+                .ValueOrDie();
+
+      // ensure that callbacks are not executed immediately on a background thread
+      gen = arrow::MakeTransferredGenerator(std::move(gen),
+                                            ::arrow::internal::GetCpuThreadPool());
+    } else {
+      gen = arrow::MakeVectorGenerator(std::move(opt_batches));
+    }
+
+    return gen;
+  }
```

Comment:
Probably we should take `Scanner::ScanBatchesUnorderedAsync` and use `arrow::MakeMapGenerator` to convert it into a generator of ExecBatch, instead of using internal utilities here and making confusing comments about threads (which users should not need to deal with).

Reply:
Either that, or `SourceNodeOptions` should implement a `MakeFromScanner`. Thoughts cc @westonpace?

Reply:
Oh, seems like I already pinged you here. @vibhatha this is really the decision. I think adding one API would fix all the comments about `gen` and `internal::Initialize`.

Reply:
cleaned up the function and removed parallel option